### PR TITLE
feat: improved error handling on the CLI

### LINKED ISSUE #1180: Better error message when not initialized
- **Version**: js-ipfs version: 0.27.7
- **Platform**: Linux frea 4.14.0-2-amd64 #1 SMP Debian 4.14.7-1 (2017-12-22) x86_64 GNU/Linux
- **Subsystem**: CLI

#### Type:
Bug

#### Severity:
Low

#### Description:

`jsipfs block put` errors with an unhandled exception when IPFS wasn't  previously initialized. Instead it should print a nice error that no repository was initialized and calling `jsipfs init` first might make sense. Example:

```
$ jsipfs block put README.md
/home/vmx/src/pl/js-ipfs/node_modules/ipfs-block-service/src/index.js:64
    this._repo.blocks.put(block, callback)
                      ^

TypeError: Cannot read property 'put' of undefined
    at BlockService.put (/home/vmx/src/pl/js-ipfs/node_modules/ipfs-block-service/src/index.js:64:23)
    at waterfall (/home/vmx/src/pl/js-ipfs/src/core/components/block.js:51:43)
    at nextTask (/home/vmx/src/pl/js-ipfs/node_modules/async/waterfall.js:16:14)
    at next (/home/vmx/src/pl/js-ipfs/node_modules/async/waterfall.js:23:9)
    at /home/vmx/src/pl/js-ipfs/node_modules/async/internal/onlyOnce.js:12:16
    at waterfall (/home/vmx/src/pl/js-ipfs/src/core/components/block.js:31:20)
    at nextTask (/home/vmx/src/pl/js-ipfs/node_modules/async/waterfall.js:16:14)
    at exports.default (/home/vmx/src/pl/js-ipfs/node_modules/async/waterfall.js:26:5)
    at Function.put.promisify (/home/vmx/src/pl/js-ipfs/src/core/components/block.js:28:7)
    at Object.put (/home/vmx/src/pl/js-ipfs/node_modules/promisify-es6/index.js:32:27)
```

#### Steps to reproduce the error:

Make sure you don't have a repository (i.e. `rm -Rf ~/.jsipfs`). Then run:

    jsipfs block put <some-file>

### PR DESCRIPTION
**Pull request by hugomrdias**
This PR should improve error handling and help messages in the cli. All messages printed to the console follow the go implementation.

In order to do this i used the yargs `fail()` method to catch and handle errors from the commands (https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/cli/bin.js#L81-L104).
Previously we were using the `parse()` method which isn't made to handle errors, its for running yargs in headless mode to build cli tests.

The `fail()` method handles promises returned from the commands handler, so its the perfect top-level spot to catch errors and output help specific to the cli. Like this:
https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/cli/bin.js#L87-L90
Another nice thing is that we can hide the stack trace behind a flag `--debug or -D`.

In the PR i keep an hybrid solution only `daemon`, `init` and `id` commands are ported to return promises and use `fail()`. Everything else works almost the same, only two things changed:
 - moved away from yargs singleton mode, now we actually have two different instances [here](https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/cli/bin.js#L43) and [here](https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/cli/bin.js#L106)
 - all errors are actually outputted correctly

The ultimate goal if everyone agrees would be to port all commands to something like the 'id' command included in the PR and start creating errors with codes like this 
```js
Object.assign(new Error('repo is not initialized yet'),
{
    code: 'ERR_REPO_NOT_INITIALIZED',
    path: node._repo.path
})
```  
  
Basically just adding error codes just like node does, so we can do this
```js
if (err.code === 'ERR_REPO_NOT_INITIALIZED') {
     return `No IPFS repo found in ${err.path}. 
please run: 'ipfs init'`
}
```

instead of this
```js
if (err.message.match(/not found/) || // indexeddb
    err.message.match(/ENOENT/) || // fs
    err.message.match(/No value/) // memory
) {
    return cb(null, false)
}
```

In the end we could delete all [this](https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/cli/bin.js#L105-L145) and do all the cli specific error handling inside `fail()`.

Needs:
- https://github.com/ipfs/js-ipfsd-ctl/pull/241

Fixes #1180

### CODE REVIEW

**Comment by diasdavid** [comment]
Needed PRs have been reviewed and merged. Not blocked anymore

**Comment by hugomrdias** [comment]
@diasdavid this PR is complete and with https://github.com/ipfs/js-ipfsd-ctl/pull/241 linked tests pass locally.

**Comment by diasdavid** [comment]
@alanshaw mind doing the last review?

**Comment by diasdavid** [comment]
I'm puzzled on why the "better CLI errors" is the fix for changing the error on ipfs-repo introduced in -- https://github.com/ipfs/js-ipfs-repo/commit/755b5c60436d1ccb58be71fc18b637b51542a107 --.

File: src/cli/commands/init.js [hunk 1]
```diff
@@ -22,33 +21,26 @@ module.exports = {
...
+    const path = getRepoPath()
 
     print(`initializing ipfs node at ${path}`)
 
-    const node = new IPFS({
+    return IPFS.createNodePromise({
       repo: new Repo(path),
       init: false,
       start: false
-    })
```
Comment (alanshaw) [inline src/cli/commands/init.js]: Does `createNodePromise` resolve when `start: false`?

File: src/cli/utils.js [hunk 2]
```diff
@@ -38,46 +38,33 @@ function getAPICtl (apiAddr) {
...
 
-exports.getIPFS = (argv, callback) => {
+exports.getNodeOrAPI = (argv, stateOptions = {forceInitialized: true}) => {
+  log('get node or api async')
   if (argv.api || isDaemonOn()) {
-    return callback(null, getAPICtl(argv.api), (cb) => cb())
+    return Promise.resolve(getAPICtl(argv.api))
   }
 
-  // Required inline to reduce startup time
```
Comment (alanshaw) [inline src/cli/utils.js]: Please leave this comment

File: src/cli/utils.js [hunk 3]
```diff
@@ -38,46 +38,33 @@ function getAPICtl (apiAddr) {
...
   const IPFS = require('../core')
-  const node = new IPFS({
+  return IPFS.createNodePromise({
     repo: exports.getRepoPath(),
     init: false,
     start: false,
```
Comment (alanshaw) [inline src/cli/utils.js]: Also here, does `createNodePromise` resolve when `start` is `false`?

File: src/cli/commands/id.js [hunk 4]
```diff
@@ -15,12 +16,11 @@ module.exports = {
...
-        throw err
-      }
-
-      print(JSON.stringify(id, '', 2))
-    })
+    return getNodeOrAPI(argv)
+      .then(node => Promise.all([Promise.resolve(node), node.id()]))
+      .then(([node, id]) => {
+        print(JSON.stringify(id, '', 2))
+        node.stop()
```
[the remark this answers is not in the document]
Comment (VictorBjelkholm) [reply src/cli/commands/id.js]: I'm not sure why this command is requiring a online node? Shouldn't this just work and return the ID without having to start a node? Feels wildly inefficient.

Looking at what go-ipfs does: when you call ID with a online daemon, it returns the ID object but with swarm address. When offline, it returns the same but with addresses empty. 

In contrast, what this command seems to be doing is to either grab the API if daemon is running but if none is running, start a node.

File: src/core/index.js [hunk 5]
```diff
@@ -138,8 +138,42 @@ class IPFS extends EventEmitter {
...
+ *
+ * The Promise waits for the Ready Event to resolve with a IPFS instance
+ * and rejects with Error Events. Rejections can be customized the with
+ * second param.
+ *
+ * @param {object} options - IPFS node options
+ * @param {object} stateOptions - Node state options to reject with error
+ * @returns {IPFS}
+ */
+IPFS.createNodePromise = (options = {}, stateOptions = {}) => {
```
[the remark this answers is not in the document]
Comment (Mr0grog) [reply src/core/index.js]: Oh, if this is going to add a new public API, it also needs to add documentation for it in the README: https://github.com/ipfs/js-ipfs/blob/master/README.md#api

File: src/cli/commands/id.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanshaw) [reply src/cli/commands/id.js]: That too, but I think that should be fixed outside of this PR.

File: src/core/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (alanshaw) [reply src/core/index.js]: Yes also!

File: src/cli/bin.js [hunk 6]
```diff
@@ -2,104 +2,161 @@
...
   pkg,
   updateCheckInterval: 1000 * 60 * 60 * 24 * 7 // 1 week
 }).notify()
 
-const args = process.argv.slice(2)
+const MSG_USAGE = `Usage:
+ipfs - Global p2p merkle-dag filesystem.
+
+  ipfs [options] <command> ...`
+  const MSG_EPILOGUE = `Use 'ipfs <command> --help' to learn more about each command.
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/cli/bin.js]: Sorry where? 26 has no identation. It's exactly the same as the go impl.

[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/cli/bin.js]: i see now, its `   const` eslint auto fixes that wrong.

[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/cli/bin.js]: fixed with last commit

File: src/cli/bin.js [hunk 7]
```diff
@@ -2,104 +2,161 @@
...
 const path = require('path')
-const utils = require('./utils')
-const print = utils.print
+const yargs = require('yargs/yargs')
+const updateNotifier = require('update-notifier')
+const readPkgUp = require('read-pkg-up')
+const { disablePrinting, print, getNodeOrAPI } = require('./utils')
+const addCmd = require('./commands/files/add')
+const catCmd = require('./commands/files/cat')
+const getCmd = require('./commands/files/get')
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/cli/bin.js]: fixed with last commit

File: src/cli/utils.js (hunk 2 above)
Comment (hugomrdias) [reply src/cli/utils.js]: fixed with last commit

File: src/core/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/core/index.js]: This is a static factory method, with the objective of creating a node if it is started or initialised depends on the options.
A startNode factory would imply `options = {start:true}`. 
I can just move this to `utils.js` and the scope would be cli only.

File: src/cli/commands/init.js (hunk 1 above)
Comment (hugomrdias) [reply src/cli/commands/init.js]: yes, ready is always emitted if an error doesn't happen 
https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/core/boot.js#L53

File: src/cli/utils.js (hunk 3 above)
Comment (hugomrdias) [reply src/cli/utils.js]: yes, ready is always emitted if an error doesn't happen
https://github.com/ipfs/js-ipfs/blob/feat/better-errors-cli/src/core/boot.js#L53

**Comment by VictorBjelkholm** [comment]
This is marked as "blocked" but it's unclear why it's blocked.

**Comment by VictorBjelkholm** [comment]
What's the progress here or why is it blocked? 

Being able to tell when a CLI command have finished running enables us to implement `ipfs-exec` for the CLI tests differently, where we can massively speed up the nodejs tests for js-ipfs.

**Comment by alanshaw** [comment]
@hugomrdias I think https://github.com/ipfs/js-ipfs/pull/1834 ended capturing a lot (if not all) of what this PR was trying to address. Shall we close this PR (it's super out of date now) or do you think there are parts of this PR that would still be useful and that aren't already address or about to be addressed by near future PRs (I'm thinking async/await endeavour)?

**Comment by daviddias** [comment]
Hi! js-ipfs master just got a whole new set of automated tests with #2528, https://github.com/ipfs/js-ipfs/pull/2440 and also running some of the test suites from our early testers (hi5 to  @achingbrain for setting it all up!). **Would you mind rebasing the master branch on this PR to ensure it runs all the latest tests? Thank you!**
